File: src/gen_1/solvers/ConstituientSolver.ts
import { CommonPoint, CommonPoints } from "../../core/CommonTypes.ts";
import { IRange } from "../../core/Range.ts";
import { ISolver, Solver } from "../../core/Solver.ts";
import { toSymmetricIndex } from "../../core/SymmetricIndex.ts";
import { Token, Word } from "../../core/Types.ts";

export class ConstituientSolver<T extends Token | Word> extends Solver<T> {
    readonly cPa: CommonPoints<T> = [];

    constructor(a: IRange<T>, b: IRange<T>) {
        super(a, b);
    }

    public override process = (): ISolver<T> => {
        const { a, b, step, cPa } = this;

        const crs = a.rs.intersection(b.rs); // +ve common (primitive) set

        const aPpam = toSymmetricIndex(a.ra, crs); // +ve
        const bPpam = toSymmetricIndex(b.ra, crs); // +ve

        const cra = [...crs]; // +ve common (primitive) array

        for (let ci = 0; ci < cra.length; ci++) {
            const value = cra[ci];

            const apa = aPpam.get(value)!;
            const bpa = bPpam.get(value)!;

            for (let ai = 0; ai < apa.length; ai++) {
                for (let bi = 0; bi < bpa.length; bi++) {
                    const ap = apa[ai];
                    const bp = bpa[bi];

                    const cp: CommonPoint<T> = [ap, bp, value];

                    cPa.push(cp);

                    step();
                }
            }
        }

        return this;
    };
}

File: src/gen_1/solvers/constituient_solver.py
from core import solver
from core.sequence import Sequence
from core.symmetric_index import to_symmetric_index
from core.commonality import CommonPoint
from core.types import T

# ConstituientSolver: structureless mass of the solution-relative general-domain


class ConstituientSolver[T](solver.AbstractSolver[T, CommonPoint]):
    def process(self, a: Sequence, b: Sequence):
        common_set = a.elements.intersection(b.elements)

        a_x_value_position_map = to_symmetric_index(a.values, common_set)
        b_x_value_position_map = to_symmetric_index(b.values, common_set)

        common_values = list[T](common_set)

        for value in common_values:
            a_positions = a_x_value_position_map.get(value)
            b_positions = b_x_value_position_map.get(value)

            if not a_positions or not b_positions:
                continue

            for ap in a_positions:
                for bp in b_positions:
                    common = CommonPoint(ap, bp, value)

                    self.add(common)

                    self.step()

        return self
